src/main/java/com/uber/profiling/AgentImpl.java

```diff
@@ -60,173 +67,190 @@ public void run(Arguments arguments, Instrumentation instrumentation, Collection
+                });
+            reporterProfiler.setIntervalMillis(metricInterval);
+
+            profilers.computeIfAbsent(StacktraceCollectorProfiler.PROFILER_NAME, k -> {
+                StacktraceCollectorProfiler stacktraceCollectorProfiler = new StacktraceCollectorProfiler(
+                    ((StacktraceReporterProfiler) reporterProfiler).getBuffer(),
+                    AgentThreadFactory.NAME_PREFIX);
+                return stacktraceCollectorProfiler;
+            }).setIntervalMillis(arguments.getSampleInterval());
         }
 
         if (arguments.isIoProfiling()) {
-            IOProfiler ioProfiler = new IOProfiler(reporter);
-            ioProfiler.setTag(tag);
-            ioProfiler.setCluster(cluster);
-            ioProfiler.setIntervalMillis(metricInterval);
-            ioProfiler.setProcessUuid(processUuid);
-            ioProfiler.setAppId(appId);
-
-            profilers.add(ioProfiler);
+            profilers.computeIfAbsent(IOProfiler.PROFILER_NAME, k -> {
+                IOProfiler ioProfiler = new IOProfiler(reporter);
+                ioProfiler.setTag(tag);
+                ioProfiler.setCluster(cluster);
+                ioProfiler.setProcessUuid(processUuid);
+                ioProfiler.setAppId(appId);
+
+                return ioProfiler;
+            }).setIntervalMillis(metricInterval);
         }
-        
-        return profilers;
     }
 
-    private void scheduleProfilers(Collection<Profiler> profilers) {
-        int threadPoolSize = Math.min(profilers.size(), MAX_THREAD_POOL_SIZE);
-        ScheduledExecutorService scheduledExecutorService = Executors.newScheduledThreadPool(threadPoolSize, new AgentThreadFactory());
+    /**
+     * Run profilers at start-up, once
+     */
+    private void runProfilers() {
+        if (started) {
+            logger.info("Profilers already started, do not run them again");
+            return;
+        }
+        for (Profiler profiler : profilers.values()) {
+            try {
+                profiler.profile();
+                logger.info("Ran profiler: " + profiler);
+            } catch (Throwable ex) {
+                logger.warn("Failed to run profiler: " + profiler, ex);
+            }
+        }
+        started = true;
+    }
 
-        for (Profiler profiler : profilers) {
+    /**
+     * (Re)schedule periodic profilers
+     */
+    private void scheduleProfilers() {
+        if (scheduledExecutorService == null) {
+            int threadPoolSize = Math.min(profilers.size(), MAX_THREAD_POOL_SIZE);
+            scheduledExecutorService = new ScheduledThreadPoolExecutor(threadPoolSize, new AgentThreadFactory());
+            scheduledExecutorService.setRemoveOnCancelPolicy(true);
+        }
+
+        for (Profiler profiler : profilers.values()) {
+            if (profiler.getScheduleHandler() != null) {
+                //cancel previous task if already scheduled
+                profiler.getScheduleHandler().cancel(false);
```

Comment:
Suggest removing getScheduleHandler() method and hide the logic of this cancelling inside the profiler. e.g. 

if (profiler.isRunning()) {
  profiler.cancel()
}

Reply:
good call - updated